File: C++/Chains/MdsOmChainName.cpp
//******************************************************************************
//
// chainname.cxx
// Implementation of chain name class
//
//******************************************************************************

static const char* fileId = "$Id: MdsOmChainName.cxx 2054 2007-09-26 19:55:21Z f012571 $";

#include "MdsOmChains.h"

#include <ctype.h>

namespace MdsOmNs {
//******************************************************************************
//
// Constructor
//
//******************************************************************************
MdsOmChainName::MdsOmChainName()
{
}

//******************************************************************************
//
// Destructor
//
//******************************************************************************
MdsOmChainName::~MdsOmChainName()
{
}

//******************************************************************************
//
// Initialise chain name object
//
//******************************************************************************
MdsOmStatusCode MdsOmChainName::init(const char *source, const char* symbol, ChainConfig *pchainConfig)
{
	if (!source || !symbol || !pchainConfig) {
		return MDS_OM_STATUS_NULL_ARG;
	}

	// Store the pointer to chain config object for later reference
	_pchainConfig = pchainConfig;
	
	this->source = source;
	this->symbol = symbol;
	this->subject = string(source) + "." + string(symbol);

	return MDS_OM_STATUS_OK;
}

//******************************************************************************
// GetNextLinkName
// Given the name of a link get the next link name.
// This is used when a publisher uses AddElement() and there are more
//	 then 14 elements, and we need to create a new link, and give it a name.
//******************************************************************************
bool MdsOmChainName::getNextLinkSubject(char *pszNextLinkSubject, int nSize)
{
	//----------------------------------------------------------------------------
	// This chain is going to be published, and publish chains have a symbol
	// naming convention as such:
	//	 n#chain_name
	// where 'n' is index of chain(ie 0,1,2...)
	//----------------------------------------------------------------------------
	const char *pszSymbol = getSymbol();
	const char* s = strchr(pszSymbol, '#');
	if (!s) {
		return false;
	}

	// Work out index of current link, ie up to # is numeric
	char szTmp[MAX_STR_TEK_SUBJECT+1];
	int i = 0;
	while (*pszSymbol != '#') {
		char c = *pszSymbol++;
		if (!isdigit(c)) return false;			// must be only number before the #
		szTmp[i++] = c;
	}
	if (i == 0) return false;					// must be at least 1 number before the #
	szTmp[i] = '\0';
	int nLinkIndex = atoi(szTmp) + 1;		// next number for the new link

	// Create symbol of next link(w/o the source)
	snprintf(pszNextLinkSubject, nSize, "%d%s", nLinkIndex, s);

	mama_log(MAMA_LOG_LEVEL_FINE, "MdsOmChainName::GetNextLinkSubject: sym=%s new=%s", getSubject(), pszNextLinkSubject);
	
	return true;
}

}

File: Java/MdsOmHelperClasses/src/com/local/mds/mercury/helper/chain/MdsOmChainName.java
//******************************************************************************
//
// chainname.cxx
// Implementation of chain name class
//
//******************************************************************************

package com.jpmorgan.mds.mercury.helper.chain;

import com.jpmorgan.mds.mercury.helper.MdsOmStatus;
import com.jpmorgan.mds.mercury.helper.MdsOmUtil;
import com.wombat.mama.Mama;
import com.wombat.mama.MamaLogLevel;

class MdsOmChainName {
	
	ChainConfig chainConfig = null;
	
	String source = "";
	String symbol = "";
	String subject = "";

	public MdsOmChainName()
	{
	}
	
	//******************************************************************************
	// Initialise chain name object
	//******************************************************************************
	public short init(String source, String symbol, ChainConfig chainConfig)
	{
		if (MdsOmUtil.isBlank(source) || MdsOmUtil.isBlank(symbol) || chainConfig == null) {
			Mama.log(MamaLogLevel.ERROR, String.format("MdsOmmChainName.init: null arg source=%s symbol=%s config=%s", source, symbol, chainConfig));
			return MdsOmStatus.NULL_ARG;
		}
	
		// Store the pointer to chain config object for later reference
		this.chainConfig = chainConfig;
		
		this.source = source;
		this.symbol = symbol;
		this.subject = source + "." + symbol;
	
		return MdsOmStatus.OK;
	}
	
	//******************************************************************************
	// GetNextLinkName
	// Given the name of a link get the next link name.
	// This is used when a publisher uses AddElement() and there are more
	//	 then 14 elements, and we need to create a new link, and give it a name.
	//******************************************************************************
	String getNextLinkSubject()
	{
		//----------------------------------------------------------------------------
		// This chain is going to be published, and publish chains have a symbol
		// naming convention as such:
		//	 n#chain_name
		// where 'n' is index of chain(ie 0,1,2...)
		//----------------------------------------------------------------------------
		int index = symbol.indexOf('#');
		if (index == -1) {
			return null;
		}
	
		// Work out index of current link, ie up to # is numeric
		String linkIndexStr = symbol.substring(0, index);
		int linkIndex = 0;
		try {
			linkIndex = Integer.valueOf(linkIndexStr).intValue();
		} catch (NumberFormatException e) {
			return null;
		}
	
		// Create symbol of next link(w/o the source)
		linkIndex++;
		String linkBaseStr = symbol.substring(index);
		String nextLinkSubject = "" + linkIndex + linkBaseStr;
	
		Mama.log(MamaLogLevel.FINE, String.format("GetNextLinkSubject: sym=%s new=%s", subject, nextLinkSubject));
		
		return nextLinkSubject;
	}
}
